### ISSUE TITLE
Issue on Unblock Friend.

### ISSUE THREAD

**Issue by krunalbabaria** [sungwoncho/has_friendship#34]
> I am phasing one issue on unblock. when i perform block at that time only status change. but when i unblock perform at that time entry remove from table instead of status change.

**Comment by sungwoncho** [comment]
When we first implemented block_friend, it was for blocking someone from making a friend request. That is why `unblock_friend` just destroys the existing friendship ([see here](https://github.com/sungwoncho/has_friendship/blob/ccd15c017e32c4cbd590542895086a27a0af9093/lib/has_friendship/friendable.rb#L75)).

But I see that it is useful to block/unblock a friend. On top of my head, a solution would be to have 2 different status for blocked friendship. Maybe `blocked_friend`, `blocked_non_friend` (see [here](https://github.com/sungwoncho/has_friendship/blob/ccd15c017e32c4cbd590542895086a27a0af9093/lib/has_friendship/friendship.rb#L4)). That way we can implement `unblock_friend` method to either destroy or restore friendship.

Let me know if you have other suggestions.

**Comment by derekcannon** [comment]
Just to jump in on this, I think that unblock_friend performing two different functions depending on the state of the blocked individual (whether he is a friend or not) is a little weird. I think it would be better to make two explicit methods for these. Thoughts?

**Comment by krtschmr** [comment]
block somebody should not rely on a friendship request
in general, i can make friendships, but out of nothing i can block friends.

`dee.block(tim)`
`tim.blocked_by?(dee) # true`

if we have been friends, and somehow a friend blocks another, then the friendships is cancelled.

**Comment by Pensarfeo** [comment]
@krtschmr : do you mean you should be able to block people even if you do not have any type of relation with them? Why would that happen withing the scope of this gem?
I agree though, the naming of the method is misleading (my bad :). Unblock friend would imply resetting back the friendship rather than destroying the relation.

**Comment by krtschmr** [comment]
basically it's a question of /the abilities of a gem/

i checked a few gems and like this most - but with lacking support of 
blocks it's useless for us (and propably many others). instead of 
hijacking a block-feature inside i propably build the whole thing from 
scratch, based on our needs.

i just wanted to let you know, that if you can block a 
friendship-request, you should be able to block people in general.

if somebody brings negative content all day, i want to block him (like 
facebook), regardless of beeing friend with him. i don't read his 
content in the future and he can't send me a friendship.




---
Diese E-Mail wurde von Avast Antivirus-Software auf Viren geprüft.
https://www.avast.com/antivirus

**Comment by Pensarfeo** [comment]
I see in which context you want to use the gem. This functionality could be added fairly easily. In the mean time you could just create a friendship and block it in the same controller.
```ruby
@mac.friend_request(@dee)
@dee.accept_request(@mac)
@mac.block_friend(@dee)
```

**Comment by krtschmr** [comment]
yes i know. that's what we are doing since pretty mich 6 years and since 
we now re-programm everything it's time to also not create a friendship 
to block it. but seems like this is the way to do in most gems.


---
Diese E-Mail wurde von Avast Antivirus-Software auf Viren geprüft.
https://www.avast.com/antivirus

**Comment by Pensarfeo** [comment]
What you want to do goes a bit outside of the scope of the gem. However I should be easy for you to add that functionality!

**Comment by krtschmr** [comment]
I come back to this.
The problem here is - if you block somebody, out of no reason, then you have a blocked friendship with him. if you unblock him you gained a friend. 
this is somehow a bad user-experience. so let's think about what would be the best way to implement that? 


      def block_somebody!(friend)
        friend_request(friend)
        block_friend(friend)
      end

this blocks anything without an active friendship, so `current_user.block_somebody!(annoying_user)`

**Comment by krtschmr** [comment]
i think a more universal use-case would be

```

  def friend_with?(user)
    # A friended B
    a.friend_with?(b) # false
    b.friend_with?(a) # false
    # B approved
    a.friend_with?(b) # true
    b.friend_with?(a) # true
  end

  def pending_friend_request?(user)
    # A friended B
    a.pending_friend_request?(b) # true
    b.pending_friend_request?(a) # true
  end

  def blocked_by?(user)
    # A Blocked B
    a.blocked_by?(b) # false
    b.blocked_by?(a) # true
  end

  def blocked?(user)
    # A Blocked B
    a.blocked?(b) # true
    b.blocked?(a) # false
  end
```

seems like i'll make this myself as hijacking work-arounds into your gem would be oversized.

**Comment by krtschmr** [comment]
specs would look like

https://gist.github.com/krtschmr/c9cb75ad364140ee5b6b953a1b1bb0cf

**Comment by stingrayzboy** [comment]
Hi All,
I have made my entire functionality based on this gem.
The fault with the blocking mechanism is its actually trying to alter the actual friendship by executing on both of them.
Granted for the Friend request there will be 2 entries.
But suppose if A and B are friends
A blocks B we shouldn't change the status of both the records to 3(blocked)
Rather the entry should be only in A so that A.blocked_friends shows B and not vice-versa
So with this approach the Unblock method will again change the status to 2 (friends)

All this while B still had A in his friend listing as B had not Blocked A
there should be A method called has_been_blocked where B will get to know if he is blocked by A
B.has_been_blocked A #=> true
this method can be used by the functionalities to implement the Block UnBlock Mechanism Properly.

**Comment by krtschmr** [comment]
@stingrayzboy  have a look at this approach https://gist.github.com/krtschmr/c9cb75ad364140ee5b6b953a1b1bb0cf
